drivers/clock_control/clock_control_renesas_cpg_mssr.c

```diff
@@ -58,3 +58,319 @@ int rcar_cpg_mstp_clock_endisable(uint32_t base_address, uint32_t module, bool e
+	uint32_t left = 0, right = data->clk_info_table_size[domain] - 1;
+
+	while (left <= right) {
+		uint32_t middle = left + ((right - left) >> 1);
+		struct cpg_clk_info_table *item = &data->clk_info_table[domain][middle];
+		uint32_t module = item->module;
+
+		if (module == id) {
+			return item;
+		}
+
+		if (module < id) {
+			left = middle + 1;
+		} else {
+			right = middle - 1;
+		}
+	}
+
+	LOG_ERR("%s: can't find clk info (domain %u module %u)", dev->name, domain, id);
+	return NULL;
+}
+
+static uint32_t rcar_cpg_get_divider(const struct device *dev, struct cpg_clk_info_table *clk_info)
+{
+	mem_addr_t reg_addr;
+	mm_reg_t reg_val;
+	uint32_t divider = RCAR_CPG_NONE;
+	struct rcar_cpg_mssr_data *data = dev->data;
+
+	if (clk_info->domain == CPG_MOD) {
+		return 1;
+	}
+
+	reg_addr = clk_info->offset;
+	if (reg_addr == RCAR_CPG_NONE) {
+		/* if we don't have valid offset, in is equal to out */
+		return 1;
+	}
+
+	reg_addr += data->base_addr;
+	reg_val = sys_read32(reg_addr);
+
+	if (data->get_div_helper) {
+		divider = data->get_div_helper(reg_val, clk_info->module);
+	}
+
+	if (!divider) {
+		return RCAR_CPG_NONE;
+	}
+
+	return divider;
+}
+
+static int rcar_cpg_update_out_freq(const struct device *dev, struct cpg_clk_info_table *clk_info)
+{
+	uint32_t divider = rcar_cpg_get_divider(dev, clk_info);
+
+	if (divider == RCAR_CPG_NONE) {
+		return -EINVAL;
+	}
+
+	clk_info->out_freq = clk_info->in_freq / divider;
+	return 0;
+}
+
+static int rcar_cpg_get_in_update_out_freq(const struct device *dev,
+					   struct cpg_clk_info_table *clk_info)
+{
+	int freq = -ENOTSUP;
+	struct cpg_clk_info_table *parent_clk;
+
+	if (!clk_info) {
+		return freq;
+	}
+
+	if (clk_info->in_freq != RCAR_CPG_NONE) {
+		if (clk_info->out_freq == RCAR_CPG_NONE) {
+			if (rcar_cpg_update_out_freq(dev, clk_info) < 0) {
+				return freq;
+			}
+		}
+		return clk_info->in_freq;
+	}
+
+	parent_clk = clk_info->parent;
+
+	freq = rcar_cpg_get_in_update_out_freq(dev, parent_clk);
+	if (freq < 0) {
+		return freq;
+	}
+
+	clk_info->in_freq = parent_clk->out_freq;
+
+	freq = rcar_cpg_update_out_freq(dev, clk_info);
+	if (freq < 0) {
+		return freq;
+	}
+
+	return clk_info->in_freq;
+}
+
+static int rcar_cpg_get_out_freq(const struct device *dev, struct cpg_clk_info_table *clk_info)
+{
+	int freq;
+
+	if (clk_info->out_freq != RCAR_CPG_NONE) {
+		return clk_info->out_freq;
+	}
+
+	freq = rcar_cpg_get_in_update_out_freq(dev, clk_info);
+	if (freq < 0) {
+		return freq;
+	}
+
+	return clk_info->out_freq;
+}
+
+bool rcar_cpg_is_pwr_of_two(uint32_t val)
+{
+	return (val & (val - 1)) ? false : true;
+}
+
+static void rcar_cpg_change_children_in_out_freq(const struct device *dev,
+						 struct cpg_clk_info_table *parent)
+{
+	struct cpg_clk_info_table *children_list = parent->children_list;
+
+	while (children_list) {
+		children_list->in_freq = parent->out_freq;
+
+		if (rcar_cpg_update_out_freq(dev, children_list) < 0) {
+			__ASSERT_NO_MSG(0);
```

Comment:
I believe that this is a critical issue and it can be in the next 4 cases:
* divider is zero (with current implementation of board specific divider helper function it is impossible);
* we don't have board specific implementation of get divider helper function;
* we don't have this module in a table (for some of call chains of this function it is impossible);
* impossible value is set in clock register divider bits.

Reply:
This list should be in the comment above that assertion. Also, how it is supposed to work in the release build?